### PR TITLE
fix(http): include transferCache when cloning HttpRequest

### LINKED ISSUE #54924: HttpRequest.clone forgets transferCache
### Which @angular/* package(s) are the source of the bug?

common

### Is this a regression?

No

### Description

When cloning the HttpRequest, like keycloak does in it's interceptor, the clone-method forgets the transferCache-property.
```
req.transferCache = true;
this.keycloak.addTokenToHeader(req.headers).pipe(mergeMap(headersWithBearer => {
      const kcReq = req.clone({
        headers: headersWithBearer
      });
      return next.handle(kcReq);
    }));
req.transferCache === true;
kcReq.transferCache === undefined;
```

The HttpRequest-constructor expects the transferCache to be set in the options-parameter, but the clone-method does not provide it.

This code should fix the problem.

`
const transferCache = update.transferCache || this.transferCache ;

    // Finally, construct the new HttpRequest using the pieces from above.
    return new HttpRequest(method, url, body, {
      params,
      headers,
      context,
      reportProgress,
      responseType,
      withCredentials,
     transferCache ,
    });`

### Please provide a link to a minimal reproduction of the bug

_No response_

### Please provide the exception or error you saw

_No response_

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 17.3.0
Node: 20.9.0
Package Manager: npm 10.1.0
OS: win32 x64

Angular: 17.3.0
... animations, cli, common, compiler, compiler-cli, core, forms
... localize, platform-browser, platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1703.0
@angular-devkit/build-angular   17.3.0
@angular-devkit/core            17.3.0
@angular-devkit/schematics      17.3.0
@schematics/angular             17.3.0
rxjs                            7.8.1
typescript                      5.3.3
zone.js                         0.14.4
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by mlz11**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Fixes #54924.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/common/http/src/request.ts [hunk 1]
```diff
@@ -479,6 +482,11 @@ export class HttpRequest<T> {
     const url = update.url || this.url;
     const responseType = update.responseType || this.responseType;
 
+    // Carefully handle the transferCache to differentiate between
+    // `false` and `undefined` in the update args.
+    const transferCache =
+      update.transferCache !== undefined ? update.transferCache : this.transferCache;
+
```
Comment (alan-agius4) [inline packages/common/http/src/request.ts:489]:
```suggestion
    // Carefully handle the transferCache to differentiate between
    // `false` and `undefined` in the update args.
    const transferCache = update.transferCache ?? this.transferCache;

```

**Review by alan-agius4 — CHANGES_REQUESTED** [review]
Thanks for this. 

Can you please update the commit message scope from `common` to `http`?

**Comment by mlz11** [comment]
> Thanks for this.
> 
> Can you please update the commit message scope from `common` to `http`?

Of course, done ✅

File: packages/common/http/src/request.ts (hunk 1 above)
Comment (mlz11) [reply packages/common/http/src/request.ts]: Done ✅

**Review by alan-agius4 — APPROVED** [review]
LGTM, thanks for your contribution

**Comment by alan-agius4** [comment]
Can you please run update the golden files via `yarn bazel run //packages/common:common_api.accept`?

**Comment by mlz11** [comment]
> Can you please run update the golden files via `yarn bazel run //packages/common:common_api.accept`?

Of course, done ✅ 
I was wondering why tests were failing. Is this the reason ?

File: packages/common/http/src/request.ts [hunk 2]
```diff
@@ -479,6 +482,10 @@ export class HttpRequest<T> {
     const url = update.url || this.url;
     const responseType = update.responseType || this.responseType;
 
+    // Carefully handle the transferCache to differentiate between
+    // `false` and `undefined` in the update args.
```
Comment (kbrilla) [inline packages/common/http/src/request.ts:486]: now it also handles `null` so mayby add that to comment : `undefined or null` here and in other places that were edited. Also why not add move this logic together with `withCredentials` and `reportProgress`? becouse now  
this comment is duplicated in lines 485 and 495

**Review by AndrewKushnir — APPROVED** [review]

**Review by alan-agius4 — APPROVED** [review]
LGTM

**Review by dylhunn — APPROVED** [review]
reviewed-for: public-api

**Review by jessicajaniuk — APPROVED** [review]
reviewed-for: public-api

**Review by atscott — APPROVED** [review]
reviewed-for: public-api

**Comment by dylhunn** [comment]
This PR was merged into the repository by commit b80434a07b57229492040cbb4932c165fab44e6c.

**Pull request closed without merging**